File: verification/scripts/r/diagnostics/test_shapiro_wilk.R
#!/usr/bin/env Rscript
# ============================================================================
# Shapiro-Wilk Test Reference Implementation (R)
# ============================================================================
# This script generates reference values for the Shapiro-Wilk test using R's
# native shapiro.test function. The Shapiro-Wilk test checks whether residuals
# are normally distributed.
#
# Source: stats package, shapiro.test function
# Reference: Shapiro & Wilk (1965), "An analysis of variance test for normality
#            (complete samples)", Biometrika, Vol. 52, pp. 591-611
#
# Usage:
#   Rscript test_shapiro_wilk.R [csv_path] [output_dir]
#   Args:
#     csv_path  - Path to CSV file (first col = response, rest = predictors)
#                 Default: ../../datasets/csv/mtcars.csv
#     output_dir- Path to output directory
#                 Default: ../../results/r
# ============================================================================

library(stats)
library(jsonlite)

# Parse command line arguments
args <- commandArgs(trailingOnly = TRUE)

# Set defaults
default_csv <- "../../datasets/csv/mtcars.csv"
default_output <- "../../results/r"

# Parse arguments
csv_path <- ifelse(length(args) >= 1, args[1], default_csv)
output_dir <- ifelse(length(args) >= 2, args[2], default_output)

# Validate CSV path
if (!file.exists(csv_path)) {
  stop(paste("CSV file not found:", csv_path))
}

# Extract dataset name from filename
dataset_name <- tools::file_path_sans_ext(basename(csv_path))

# Read CSV data
data <- read.csv(csv_path)

# Get the column names
cols <- colnames(data)

# First column is y (dependent variable), rest are x variables (predictors)
y <- data[[1]]
if (ncol(data) > 1) {
  x_vars <- data[, -1, drop = FALSE]
} else {
  # No predictors - simple mean model
  x_vars <- NULL
}

# Fit OLS model and compute residuals
if (is.null(x_vars)) {
  # Simple mean model
  residuals <- y - mean(y)
} else {
  # OLS regression
  model <- lm(y ~ ., data = x_vars)
  residuals <- resid(model)
}

# Run Shapiro-Wilk test using R's native function
# H0: Residuals are normally distributed
# H1: Residuals are not normally distributed
sw_result <- shapiro.test(residuals)

# Extract results
w_statistic <- sw_result$statistic
p_value <- sw_result$p.value

# Determine if test passed (null hypothesis not rejected at alpha = 0.05)
alpha <- 0.05
passed <- p_value > alpha

# Create interpretation text
if (passed) {
  interpretation <- sprintf(
    "p-value = %.4f is greater than %.2f. Cannot reject H0. No significant evidence that residuals deviate from normality.",
    p_value, alpha
  )
  guidance <- "The normality assumption appears to be met. Shapiro-Wilk test does not detect significant deviation from normal distribution."
} else {
  interpretation <- sprintf(
    "p-value = %.4f is less than or equal to %.2f. Reject H0. Significant evidence that residuals deviate from normality.",
    p_value, alpha
  )
  guidance <- "Consider transforming the dependent variable (e.g., log, Box-Cox transformation), using robust standard errors, or applying a different estimation method."
}

# Create output JSON object (using array format like other diagnostic tests)
output <- list(
  test_name = "Shapiro-Wilk Test for Normality",
  statistic = c(w_statistic),
  p_value = c(p_value),
  passed = c(passed),
  interpretation = interpretation,
  guidance = guidance
)

# Create output directory if it doesn't exist
if (!dir.exists(output_dir)) {
  dir.create(output_dir, recursive = TRUE)
}

# Save to JSON with naming convention: {dataset}_shapiro_wilk.json
output_file <- file.path(output_dir, paste0(dataset_name, "_shapiro_wilk.json"))

# Write output to JSON file
write_json(output, output_file, pretty = TRUE, digits = 22)

cat("Results saved to:", output_file, "\n")

File: verification/scripts/python/diagnostics/test_shapiro_wilk.py
#!/usr/bin/env python3
# ============================================================================
# Shapiro-Wilk Test Validation Script (Python)
# ============================================================================
#
# This script runs the Shapiro-Wilk test on regression residuals using
# scipy's shapiro function and outputs the results to JSON for validation
# against the Rust implementation.
#
# Usage: python test_shapiro_wilk.py --csv <csv_file>
# Example: python test_shapiro_wilk.py --csv ../../datasets/csv/mtcars.csv
#
# Output: JSON file with test results

import argparse
import json
import os
import sys
import numpy as np
import pandas as pd
from scipy import stats

def convert_categorical_to_numeric(df, dataset_name):
    """Convert categorical columns to numeric representations."""
    non_numeric_cols = df.select_dtypes(exclude=[np.number]).columns.tolist()

    if non_numeric_cols:
        print(f"INFO: Dataset '{dataset_name}' contains non-numeric columns: "
              f"{', '.join(non_numeric_cols)}")
        print("Converting categorical variables to numeric representations...")

        for col in non_numeric_cols:
            # Use factorize for reliable categorical encoding
            df[col], uniques = pd.factorize(df[col])
            print(f"  {col}: {len(uniques)} unique values -> integer level encoding")

    return df

def parse_args():
    parser = argparse.ArgumentParser(description='Shapiro-Wilk test validation')
    parser.add_argument('--csv', required=True, help='Path to CSV file')
    parser.add_argument('--output-dir', default='../../results/python', help='Path to output directory')
    return parser.parse_args()

def main():
    args = parse_args()
    csv_file = args.csv

    # Check if file exists
    if not os.path.exists(csv_file):
        print(f"Error: File not found: {csv_file}", file=sys.stderr)
        sys.exit(1)

    # Get dataset name
    dataset_name = os.path.splitext(os.path.basename(csv_file))[0]

    # Read CSV data using pandas to handle categorical columns
    data = pd.read_csv(csv_file)

    # Convert categorical columns to numeric
    data = convert_categorical_to_numeric(data, dataset_name)

    # Convert to numpy array
    data = data.values

    # First column is y (dependent variable), rest are x variables (predictors)
    y = data[:, 0]
    if data.shape[1] > 1:
        x_vars = data[:, 1:]
    else:
        # No predictors - simple mean model
        x_vars = None

    # Fit OLS model and compute residuals
    if x_vars is None:
        # Simple mean model
        residuals = y - np.mean(y)
    else:
        # OLS regression
        # Add intercept column
        X = np.column_stack([np.ones(len(y)), x_vars])

        # Solve normal equations: beta = (X'X)^(-1) X'y
        try:
            beta = np.linalg.solve(X.T @ X, X.T @ y)
        except np.linalg.LinAlgError:
            # Use least squares if singular
            beta = np.linalg.lstsq(X, y, rcond=None)[0]

        # Compute residuals
        residuals = y - X @ beta

    # Run Shapiro-Wilk test using scipy's native function
    # H0: Residuals are normally distributed
    # H1: Residuals are not normally distributed
    w_statistic, p_value = stats.shapiro(residuals)

    # Determine if test passed (null hypothesis not rejected at alpha = 0.05)
    alpha = 0.05
    passed = p_value > alpha

    # Create interpretation text
    if passed:
        interpretation = (
            f"p-value = {p_value:.4f} is greater than {alpha:.2f}. "
            f"Cannot reject H0. No significant evidence that residuals deviate from normality."
        )
        guidance = ("The normality assumption appears to be met. Shapiro-Wilk test does not detect "
                    "significant deviation from normal distribution.")
    else:
        interpretation = (
            f"p-value = {p_value:.4f} is less than or equal to {alpha:.2f}. "
            f"Reject H0. Significant evidence that residuals deviate from normality."
        )
        guidance = ("Consider transforming the dependent variable (e.g., log, Box-Cox transformation), "
                    "using robust standard errors, or applying a different estimation method.")

    # Create output dictionary
    output = {
        "test_name": "Shapiro-Wilk Test for Normality",
        "statistic": float(w_statistic),
        "p_value": float(p_value),
        "is_passed": bool(passed),
        "interpretation": interpretation,
        "guidance": guidance
    }

    # Get output file name from CSV file name
    basename = os.path.splitext(os.path.basename(csv_file))[0]
    output_file = os.path.join(args.output_dir, f"{basename}_shapiro_wilk.json")

    # Create output directory if it doesn't exist
    if not os.path.exists(args.output_dir):
        os.makedirs(args.output_dir, exist_ok=True)

    # Write output to JSON file
    with open(output_file, 'w') as f:
        json.dump(output, f, indent=2)

    print(f"Results written to: {output_file}")
    print(f"W statistic: {w_statistic}")
    print(f"p-value: {p_value}")

if __name__ == '__main__':
    main()
